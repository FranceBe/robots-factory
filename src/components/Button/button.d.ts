import { ButtonHTMLAttributes } from 'react'

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  buttonType?: 'primary' | 'secondary'
  isActive?: boolean
  tooltipText?: string | JSX.Element
  uniqueIndex?: string
}
