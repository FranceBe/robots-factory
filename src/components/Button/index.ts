import ButtonDefault, { Button } from 'components/Button/Button.component'

export { Button }

export default ButtonDefault
