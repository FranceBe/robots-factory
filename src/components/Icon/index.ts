import { FAIconByIconProps, Icon } from 'components/Icon/Icon.component'

export { FAIconByIconProps, Icon }
