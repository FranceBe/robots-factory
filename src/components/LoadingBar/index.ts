import { LoadingBar } from 'components/LoadingBar/LoadingBar.component'

export { LoadingBar }
