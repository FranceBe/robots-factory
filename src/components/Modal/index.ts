import ModalDefault, { Modal } from 'components/Modal/Modal.component'

export { Modal }

export default ModalDefault
