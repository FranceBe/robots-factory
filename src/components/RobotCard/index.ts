import { RobotCard } from 'components/RobotCard/RobotCard.component'

export { RobotCard }
