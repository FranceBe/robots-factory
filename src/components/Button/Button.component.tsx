import { ButtonContainer, ContentContainer } from 'components/Button/button.style'
import { ButtonProps } from 'components/Button/types/button'
import { Icon } from 'components/Icon'
import React from 'react'
import ReactTooltip from 'react-tooltip'
import { palette } from 'styles/variables'

export const Button: React.FC<ButtonProps> = ({
  buttonType = 'primary',
  isActive = false,
  tooltipText,
  uniqueIndex = '1',
  ...rest
}) => {
  return (
    <div
      data-tip={tooltipText}
      data-html={true}
      data-for={`button-tooltip${uniqueIndex}`}
      data-testid={'button-container'}
    >
      <ButtonContainer buttontype={buttonType} {...rest}>
        <ContentContainer>{rest.children}</ContentContainer>
        {isActive && <Icon type={'robot'} />}
        <ReactTooltip
          html={true}
          backgroundColor={palette.white}
          textColor={palette.grey_primary}
          id={`button-tooltip${uniqueIndex}`}
        />
      </ButtonContainer>
    </div>
  )
}

export default Button
