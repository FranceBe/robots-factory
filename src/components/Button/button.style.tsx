import { ButtonProps } from 'components/Button/button'
import React from 'react'
import styled from 'styled-components'
import { fontSizes, palette, spaces } from 'styles/variables'
import { ThemeType } from 'utils/common.variables'

export const ButtonContainer = styled((props) => <button {...props} />)<ButtonProps>`
  height: ${spaces.xx_large};
  border: none;
  background-color: ${(props) =>
    props.buttontype === ThemeType.primary ? palette.pink_primary : palette.blue_primary};
  padding: ${spaces.small};
  color: ${palette.white};
  font-size: ${fontSizes.regular};
  border-radius: ${spaces.x_small};
  cursor: pointer;
  min-width: 184px;
  :hover {
    background-color: ${(props) =>
      props.buttontype === ThemeType.primary ? palette.pink_tertiary : palette.blue_tertiary};
  }
  :disabled,
  :disabled:hover {
    color: ${palette.grey_quaternary};
    background-color: ${(props) =>
      props.buttontype === ThemeType.primary ? palette.pink_secondary : palette.blue_secondary};
    cursor: not-allowed;
  }
  svg {
    position: relative;
    bottom: 15px;
    left: 75px;
    filter: drop-shadow(0 0 ${spaces.x_small} ${palette.grey_primary});
  }
`
export const ContentContainer = styled.div`
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
`
