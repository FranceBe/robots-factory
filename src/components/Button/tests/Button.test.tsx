import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Button } from 'components/Button'
import { ButtonProps } from 'components/Button/button'
import React from 'react'
import { palette } from 'styles/variables'
import { ThemeType } from 'utils/common.variables'

describe('Button component', () => {
  const initTest = (props?: ButtonProps) => render(<Button {...props}>Button text</Button>)

  describe('Common behaviour', () => {
    it('should match snapshot', () => {
      const { container } = initTest()
      expect(container.firstChild).toMatchSnapshot()
    })
    it('should have primary button style by default', () => {
      initTest()
      const button = screen.getByRole('button', { name: 'Button text' })

      expect(button).toHaveStyleRule('background-color', `${palette.pink_primary}`)
      expect(button).toHaveStyleRule('color', `${palette.white}`)
    })
    it('should display the button text', () => {
      initTest()

      expect(screen.getByText('Button text')).toBeInTheDocument()
    })
    it('should call provided function on click', () => {
      const mockedFn = jest.fn()
      initTest({ onClick: mockedFn })

      userEvent.click(screen.getByRole('button'))
      expect(mockedFn).toHaveBeenCalledTimes(1)
    })
    it('should not display a robot icon by default', () => {
      initTest()

      const robotIcon = screen.queryByRole('img', { hidden: true })
      expect(robotIcon).not.toBeInTheDocument()
    })
    it('should display a robot icon when button isActive', () => {
      initTest({ isActive: true })

      const robotIcon = screen.getByRole('img', { hidden: true })
      expect(robotIcon).toBeInTheDocument()
      expect(robotIcon).toHaveClass('fa-robot')
    })
    it('should display a tooltip when tooltipText is provided', async () => {
      initTest({ tooltipText: 'Tooltip text' })

      userEvent.hover(screen.getByTestId('button-container'))

      await waitFor(() => {
        expect(screen.getByText('Tooltip text')).toBeInTheDocument()
      })
    })
  })

  describe('Given buttonType is "primary"', () => {
    it('should have primary button style', () => {
      initTest({ buttonType: ThemeType.primary })
      const button = screen.getByRole('button', { name: 'Button text' })

      expect(button).toHaveStyleRule('background-color', `${palette.pink_primary}`)
      expect(button).toHaveStyleRule('color', `${palette.white}`)
      expect(button).toHaveStyleRule('background-color', `${palette.pink_tertiary}`, {
        modifier: ':hover',
      })
      expect(button).toHaveStyleRule('background-color', `${palette.pink_secondary}`, {
        modifier: ':disabled',
      })
      expect(button).toHaveStyleRule('color', `${palette.grey_quaternary}`, {
        modifier: ':disabled',
      })
    })
  })

  describe('Given buttonType is "secondary"', () => {
    it('should have secondary button style', () => {
      initTest({ buttonType: ThemeType.secondary })
      const button = screen.getByRole('button', { name: 'Button text' })

      expect(button).toHaveStyleRule('background-color', `${palette.blue_primary}`)
      expect(button).toHaveStyleRule('color', `${palette.white}`)
      expect(button).toHaveStyleRule('background-color', `${palette.blue_tertiary}`, {
        modifier: ':hover',
      })
      expect(button).toHaveStyleRule('background-color', `${palette.blue_secondary}`, {
        modifier: ':disabled',
      })
      expect(button).toHaveStyleRule('color', `${palette.grey_quaternary}`, {
        modifier: ':disabled',
      })
    })
  })
})
