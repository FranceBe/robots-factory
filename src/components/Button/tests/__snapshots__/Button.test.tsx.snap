// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Common behaviour should match snapshot 1`] = `
.c0 {
  height: 64px;
  border: none;
  background-color: #DB93B0;
  padding: 8px;
  color: #FFF;
  font-size: 1rem;
  border-radius: 4px;
  cursor: pointer;
  min-width: 184px;
}

.c0:hover {
  background-color: #E6B3C8;
}

.c0:disabled,
.c0:disabled:hover {
  color: #E5E5E5;
  background-color: #99677B;
  cursor: not-allowed;
}

.c0 svg {
  position: relative;
  bottom: 15px;
  left: 75px;
  -webkit-filter: drop-shadow(0 0 4px #6A6C71);
  filter: drop-shadow(0 0 4px #6A6C71);
}

.c1 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div
  data-for="button-tooltip1"
  data-html="true"
  data-testid="button-container"
>
  <button
    buttontype="primary"
    class="c0"
  >
    <div
      class="c1"
    >
      Button text
    </div>
    <div
      class="__react_component_tooltip ta7448cad-1864-4ed9-95fe-e9006661c33d place-top type-dark"
      data-id="tooltip"
      id="button-tooltip1"
    >
      <style
        aria-hidden="true"
      >
        
  	.ta7448cad-1864-4ed9-95fe-e9006661c33d {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.ta7448cad-1864-4ed9-95fe-e9006661c33d.place-top {
        margin-top: -10px;
    }
    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-top::before {
        border-top: 8px solid transparent;
    }
    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-bottom {
        margin-top: 10px;
    }
    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-left {
        margin-left: -10px;
    }
    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-left::before {
        border-left: 8px solid transparent;
    }
    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-right {
        margin-left: 10px;
    }
    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-right::before {
        border-right: 8px solid transparent;
    }
    .ta7448cad-1864-4ed9-95fe-e9006661c33d.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  
      </style>
    </div>
  </button>
</div>
`;
